### PR TITLE
Add support for filter spec

### PR DESCRIPTION
**Pull request by bishwajit-db**
Add support for filter spec.
Add Title and Description for Filters based on filter spec.
Add support for DateRange filter. 
Add support for Filters linked with multiple visualization widgets.

Resolves: #246, #252

### CODE REVIEW

File: src/databricks/labs/lsql/dashboards.py [hunk 1]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
         return spec
 
 
+@unique
+class FilterType(str, Enum):
+    """The filter widget type"""
+
+    DATE_RANGE_PICKER = "DATE_RANGE_PICKER"
+    MULTI_SELECT = "MULTI_SELECT"
+    NONE = "NONE"
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:758]: When is `None` used?

File: src/databricks/labs/lsql/dashboards.py [hunk 2]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+        Raises:
+            ValueError : If the tile is invalid.
+        """
+        super().validate()
+        if not self.metadata.is_filter():
+            raise ValueError(f"Tile is not a filter file: {self}")
+
+    def get_layouts(self) -> Iterable[Layout]:
+        """Get the layout(s) reflecting this tile in the dashboard."""
+        if not self.dataset_columns:
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:787]: I think this should be part of the `validate`

File: src/databricks/labs/lsql/dashboards.py [hunk 3]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+            logger.warning(f"Filter tile for path {self.metadata.path} has no matching dataset columns")
+            return
+        config = json.loads(self.content)
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:790]: Should be part of the `FilterHandler` see `_parse_header` on other Handlers as example

File: src/databricks/labs/lsql/dashboards.py [hunk 4]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+        filter_type = FilterType[config.get("type", "MULTI_SELECT").upper()]
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:791]: `FilterType` conversion from string should happen in the `FilterHandler` during parsing

``` python
        parser.add_argument(
            "--type",
            type=lambda v: WidgetType(v.upper()),
            default=WidgetType.AUTO,
            help=(
                "The widget type to use, see classes with WidgetSpec as parent class in "
                "databricks.labs.lsql.lakeview.model."
            ),
            dest="widget_type",
        )
```

File: src/databricks/labs/lsql/dashboards.py [hunk 5]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+        queries = []
+        for idx, (column, dataset_name) in enumerate(self.dataset_columns):
+            filter_fields = [
+                Field(name=column, expression=f"`{column}`"),
+                Field(name=f"{column}_associativity", expression="COUNT_IF(`associative_filter_predicate_group`)"),
+            ]
+            filter_query = Query(dataset_name=dataset_name, fields=filter_fields, disaggregated=False)
+            filter_named_query = NamedQuery(name=f"{filter_type}_{column}_{idx}", query=filter_query)
+            queries.append(filter_named_query)
+            control_encodings.append(ControlFieldEncoding(column, filter_named_query.name, display_name=column))
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:802]: could you move `ControlFrieldEncoding` to a variable first before appending it the list, like the `NamedQuery` above?

File: src/databricks/labs/lsql/dashboards.py [hunk 6]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+        filter_type = FilterType[config.get("type", "MULTI_SELECT").upper()]
+        control_encodings: list[ControlEncoding] = []
+        queries = []
+        for idx, (column, dataset_name) in enumerate(self.dataset_columns):
+            filter_fields = [
+                Field(name=column, expression=f"`{column}`"),
+                Field(name=f"{column}_associativity", expression="COUNT_IF(`associative_filter_predicate_group`)"),
+            ]
+            filter_query = Query(dataset_name=dataset_name, fields=filter_fields, disaggregated=False)
+            filter_named_query = NamedQuery(name=f"{filter_type}_{column}_{idx}", query=filter_query)
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:800]: Assuming columns are unique, is the `idx` needed?

File: src/databricks/labs/lsql/dashboards.py [hunk 7]
```diff
@@ -724,6 +749,72 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
...
+
+        frame = WidgetFrameSpec(
+            title=self.metadata.title,
+            show_title=len(self.metadata.title) > 0,
+            description=self.metadata.description,
+            show_description=len(self.metadata.description) > 0,
+        )
+        control_encoding_map = ControlEncodingMap(control_encodings)
+        filter_spec = filter_type.as_widget_spec()
+        spec = filter_spec(encodings=control_encoding_map, frame=frame)
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:812]: We should test if these are the right encoding for all the filter types

File: src/databricks/labs/lsql/dashboards.py [hunk 8]
```diff
@@ -893,6 +986,42 @@ def _from_dashboard_folder(cls, folder: Path) -> "DashboardMetadata":
             tiles.append(tile)
         return cls(display_name=folder.name, _tiles=tiles)
 
+    @classmethod
+    def _filter_tiles_from_folder(cls, folder: Path, query_tiles: Iterable[QueryTile]) -> list[FilterTile]:
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:990]: Move this logic into `from_tile_metadata`

File: src/databricks/labs/lsql/dashboards.py [hunk 9]
```diff
@@ -893,6 +986,42 @@ def _from_dashboard_folder(cls, folder: Path) -> "DashboardMetadata":
...
+
+        filter_tiles = []
+        for path in folder.iterdir():
+            if not path.name.endswith(".filter.json"):
+                continue
+            logger.debug(f"Loading filter spec from {path}")
+            filter_metadata = json.loads(path.read_text())
+            column = filter_metadata.get("column")
+            columns = filter_metadata.get("columns")
+            if not column and not columns:
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:1005]: Should be part of `FilterTile.validate`

File: src/databricks/labs/lsql/dashboards.py [hunk 10]
```diff
@@ -893,6 +986,42 @@ def _from_dashboard_folder(cls, folder: Path) -> "DashboardMetadata":
             tiles.append(tile)
         return cls(display_name=folder.name, _tiles=tiles)
 
+    @classmethod
+    def _filter_tiles_from_folder(cls, folder: Path, query_tiles: Iterable[QueryTile]) -> list[FilterTile]:
+        queries_from_tiles: list[NamedQuery] = []
+        for tile in query_tiles:
```
Comment (JCZuurmond) [inline src/databricks/labs/lsql/dashboards.py:992]: This flow is different then before:
- Existing: All tiles are treated separately
- New: The filter tile(s) need the queries from the other tiles.

I think we should move this to `get_layouts` where we pass `datasets` to the `get_layouts` of each `Tile`. In the `Tile` we do nothing with the datasets by default, but we use it in the `FilterTile` to build its layouts

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
@bishwajit-db : Thanks for your contribution! This first iteration functionally looks good. I gave some feedback to move the pieces to the right places. Also, could you add an integration test with a dashboard containing a filter and add a screenshot of the dashboard in the PR?

File: src/databricks/labs/lsql/dashboards.py [hunk 11]
```diff
@@ -189,6 +190,23 @@ def split(self) -> tuple[str, str]:
...
 
+class FilterHandler(BaseHandler):
+    """Handle filter files."""
+
+    def _parse_header(self, header: str) -> dict:
+        if not header:
+            return {}
+        metadata = json.loads(header)
+        metadata.pop("column", None)
+        metadata.pop("columns", None)
```
[the remark this answers is not in the document]
Comment (bishwajit-db) [reply src/databricks/labs/lsql/dashboards.py:201]: I am not aware. This was present in the spec, so I handled it here. In the implementation, it is being handled as a list with one column.

**Comment by bishwajit-db** [comment]
Closing this PR and moving to https://github.com/databrickslabs/lsql/pull/276

**Pull request closed without merging**